### PR TITLE
Support for masking sensitive debug data

### PR DESCRIPTION
**Pull request by pixobit**
**Description**
Sometimes we want to hide sensitive information from the debug trace, even if it's only for the development environment. 
This makes it possible.
To give more context to this matter, here's a link that addresses a similar issue on laravel: https://github.com/laravel/ideas/issues/1236

**Checklist:**
- [ ] Securely signed commits
- [ ] Component(s) with PHPdocs
- [ ] Unit testing, with >80% coverage
- [ ] User guide updated
- [x] Conforms to style guide

### CODE REVIEW

**Review by paulbalandan — CHANGES_REQUESTED** [review]
Please follow the coding style and make sure code is compatible with 7.3+. You're using `str_ends_with` of php8

File: system/Debug/Exceptions.php [hunk 1]
```diff
@@ -293,16 +293,61 @@ protected function render(Throwable $exception, int $statusCode)
 	 */
 	protected function collectVars(Throwable $exception, int $statusCode): array
 	{
+		$trace = $exception->getTrace();
+		if(!empty($this->config->sensitiveDataInTrace)) {
+			$this->maskSensitiveData($trace, $this->config->sensitiveDataInTrace);
+		}
```
Comment (paulbalandan) [inline system/Debug/Exceptions.php:299]:
```suggestion
		$trace = $exception->getTrace();

		if (! empty($this->config->sensitiveDataInTrace))
		{
			$this->maskSensitiveData($trace, $this->config->sensitiveDataInTrace);
		}
```

File: system/Debug/Exceptions.php [hunk 2]
```diff
@@ -293,16 +293,61 @@ protected function render(Throwable $exception, int $statusCode)
...
 	}
+	
+	/**
+	 * Mask sensitive data in the trace.
+	 *
+	 * @param array  $trace
+	 * @param array  $keysToMask
+	 * @param string $path
+	 */
+	protected function maskSensitiveData(&$trace, $keysToMask, $path = '')
```
Comment (paulbalandan) [inline system/Debug/Exceptions.php:319]:
```suggestion
	protected function maskSensitiveData(array &$trace, array $keysToMask, string $path = '') 
```

File: system/Debug/Exceptions.php [hunk 3]
```diff
@@ -293,16 +293,61 @@ protected function render(Throwable $exception, int $statusCode)
...
+				if (is_array($trace) && array_key_exists($idx, $trace)) 
+				{
+					$trace[$idx] = '******************';
+				} 
+				else if (is_object($trace) && property_exists($trace, $idx) && isset($trace->$idx)) 
+				{
+					$trace->$idx = '******************';
+				}
+			}
+		}
```
Comment (paulbalandan) [inline system/Debug/Exceptions.php:336]:
```suggestion
		foreach ($keysToMask as $keyToMask) 
		{
			$explode = explode('/', $keyToMask);
			$index   = end($explode);

			if (strpos(strrev($path . '/' . $index), strrev($keyToMask)) === 0)
			{
				if (is_array($trace) && array_key_exists($index, $trace)) 
				{
					$trace[$index] = '******************';
				} 
				elseif (is_object($trace) && property_exists($trace, $index) && isset($trace->$index)) 
				{
					$trace->$index = '******************';
				}
			}
		}
```

File: system/Debug/Exceptions.php [hunk 4]
```diff
@@ -293,16 +293,61 @@ protected function render(Throwable $exception, int $statusCode)
...
+			$trace = get_object_vars($trace);
+		}
+		
+		if(is_iterable($trace)) 
+		{
+			foreach ($trace as $pathKey => $subarray) 
+			{
+				$this->maskSensitiveData($subarray, $keysToMask, $path . '/' . $pathKey);
+			}
+		}
```
Comment (paulbalandan) [inline system/Debug/Exceptions.php:349]:
```suggestion
		if (! is_iterable($trace) && is_object($trace)) 
		{
			$trace = get_object_vars($trace);
		}
		
		if (is_iterable($trace)) 
		{
			foreach ($trace as $pathKey => $subarray) 
			{
				$this->maskSensitiveData($subarray, $keysToMask, $path . '/' . $pathKey);
			}
		}
```

**Review by paulbalandan — COMMENTED** [review]
Can you show a sample output using this ?

File: system/Debug/Exceptions.php (hunk 2 above)
Comment (pixobit) [reply system/Debug/Exceptions.php:321]: Setting $keysToMask, and $path to string type is fine, but $trace can either be an array, or object

Comment (pixobit) [reply system/Debug/Exceptions.php:321]: > Can you show a sample output using this ?

Putting the whole trace would be huge, but here's a part of it for masking 'app/settings'

 [app] => App\Models\SystemModel Object
        (
            [settings] => ******************
            [currencies] => Array
                (
                )

            [currencyFormatter] => 
            [table:protected] => 
            [primaryKey:protected] => id
            [useAutoIncrement:protected] => 1
            [builder:protected] => 
            [tempData:protected] => Array
                (
                )

            [escape:protected] => Array
                (
                )

Comment (paulbalandan) [reply system/Debug/Exceptions.php:321]: > Setting $keysToMask, and $path to string type is fine, but $trace can either be an array, or object

OK. So put `array|object` in the docblock

**Review by paulbalandan — APPROVED** [review]

**Review by paulbalandan — APPROVED** [review]

File: app/Config/Exceptions.php [hunk 5]
```diff
@@ -45,4 +45,16 @@ class Exceptions extends BaseConfig
...
 	 */
 	public $errorViewPath = APPPATH . 'Views/errors';
+        
+	/**
+	 * --------------------------------------------------------------------------
+	 * HIDE FROM DEBUG TRACE
+	 * --------------------------------------------------------------------------
+	 * Any data that you would like to hide from the debug trace.
+	 * In order to specify 2 levels, use "/" to separate.
+         * ex. ['server', 'setup/password', 'secret_token']
```
Comment (kenjis) [inline app/Config/Exceptions.php:55]: What is `'server'` in the example? What is *level* ?
What does this feature hide?
Does this feature really work?

**Pull request merged**